src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -16,28 +16,51 @@
  */
 
 #include <app/clusters/bindings/BindingManager.h>
+#include <app/clusters/bindings/bindings.h>
 #include <app/util/binding-table.h>
 #include <credentials/FabricTable.h>
+#include <lib/support/CHIPMem.h>
 
 namespace {
 
 class BindingFabricTableDelegate : public chip::FabricTableDelegate
 {
     void OnFabricDeletedFromStorage(chip::CompressedFabricId compressedFabricId, chip::FabricIndex fabricIndex)
     {
-        auto iter = chip::BindingTable::GetInstance().begin();
-        while (iter != chip::BindingTable::GetInstance().end())
+        chip::BindingTable & bindingTable = chip::BindingTable::GetInstance();
+        auto iter                         = bindingTable.begin();
+        while (iter != bindingTable.end())
         {
             if (iter->fabricIndex == fabricIndex)
             {
-                iter = chip::BindingTable::GetInstance().RemoveAt(iter);
+                iter = bindingTable.RemoveAt(iter);
             }
             else
             {
                 ++iter;
             }
         }
         chip::BindingManager::GetInstance().FabricRemoved(compressedFabricId, fabricIndex);
+
+        EmberBindingTableEntry * table = static_cast<EmberBindingTableEntry *>(
+            chip::Platform::MemoryAlloc(bindingTable.Size() * sizeof(EmberBindingTableEntry)));
+        uint8_t idx = 0;
+        if (table == nullptr)
+        {
+            return;
+        }
+        for (const EmberBindingTableEntry & entry : chip::BindingTable::GetInstance())
+        {
+            table[idx++] = entry;
+        }
+        CHIP_ERROR error = chip::Server::GetInstance().GetPersistentStorage().SyncSetKeyValue(
```

Comment:
We cannot do this directly. Please follow in the footsteps of GroupDataProvider.h/AccessControl.h in terms of interface, and the related implementations:

- Create an interface for storage of a data domain (here, binding table) that can be implemented differently by some applications
- Have all binding data model handling go through that interface, implemented outside the cluster server files
- Have ember and IM code that deals with bindings access their data through that
- Have the bindings cluster server also access its data through that interface

The default implementations of the interface (see GroupDataProviderImpl.cpp and ExampleAccessControlDelegate.cpp for examples) have full unit tests using the `TestPersistentStorageProvider`.

Going through such an interface has the following advantages:

- Different products can build/link a different implementation of the core feature including different storage mechanisms
- The feature is unit testable because not "glued" to cluster logic that has too many non-injectable complex dependencies to unit tests
- It reduces reliance on the `Server` God object
- The pattern using both iterators and indexed access for data access makes it possible to properly implement fabric-scoped access. The bindings cluster is fabric scoped.

Reply:
Re-written the code to use dta model TLVs to store the binding table. Unit tests also added. I didn't create an interface for storage since the code heavily relies on the implementation detail of the BindingTable class. 

The storage implementation requires the binding to be a linked list so that the changes to written to storage is local enough for each `Add` and `Remove`. Local changes ensure that we can roll back if the storage fails. Unit test cases are also created to verify robustness against storage failure.

If we create an interface for storage, it will be like:
```
class BindingStorageDelegate {
public:
    CHIP_ERROR StoreListHead(uint8_t headIndex);
    CHIP_ERROR StoreListItem(const EmberBindingEntry &entry, uint8_t itemIndex, uint8_t nextIndex);
    CHIP_ERROR LoadFromStorage();
};
```
This doesn't look like an good interface to me. It's too tightly bound to implementation details. I chose to implement persistent storage inside the `BindingTable` class as a result. Do we still perfer to create such an interface?